# Assignment 4: Data Wrangling and Classes in Python

## POP77001 Computer Programming for Social Scientists

## Before submission

- Do not use external libraries unless explicitly specified.
- Make sure that you can run all cells without errors.
- You can do it by clicking `Kernel`, `Restart & Run All` in the menu above.
- Make sure that you save the output by pressing Command+S / CTRL+S.
- Rename the file from `04_assignment.ipynb` to `04_assignment_studentnumber_firstname_lastname.ipynb`.
- Use Firefox browser for submitting your Jupyter notebook on Blackboard.

## Exercise 1: Data transformation

For exercises 1-3 we will be working with the [Titanic](https://github.com/pandas-dev/pandas/blob/master/doc/data/titanic.csv) dataset. It contains the list of passengers on Titanic with their sociodemographic characteristics, names and ticket details. First, create two new columns, called 'Last Name' and 'First Name' containing last name and first name of each passenger. Next, calculate the frequencies of passengers' last names. What were the 3 most common last names of passengers on the ship?

External libraries:
- `numpy`, `pandas`

In [873]:
import pandas as pd

In [874]:
titanic = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv")

In [875]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [876]:
# Exercise 1:

# Your code goes here

titanic_name = titanic['Name'] 

titanic_name.head()



0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

In [877]:
titanic_name2 = [N.replace("Mr.", "").replace("Miss", "").replace("Mrs.", "")for N in titanic_name]
titanic_name2

['Braund,  Owen Harris',
 'Cumings,  John Bradley (Florence Briggs Thayer)',
 'Heikkinen, . Laina',
 'Futrelle,  Jacques Heath (Lily May Peel)',
 'Allen,  William Henry',
 'Moran,  James',
 'McCarthy,  Timothy J',
 'Palsson, Master. Gosta Leonard',
 'Johnson,  Oscar W (Elisabeth Vilhelmina Berg)',
 'Nasser,  Nicholas (Adele Achem)',
 'Sandstrom, . Marguerite Rut',
 'Bonnell, . Elizabeth',
 'Saundercock,  William Henry',
 'Andersson,  Anders Johan',
 'Vestrom, . Hulda Amanda Adolfina',
 'Hewlett,  (Mary D Kingcome) ',
 'Rice, Master. Eugene',
 'Williams,  Charles Eugene',
 'Vander Planke,  Julius (Emelia Maria Vandemoortele)',
 'Masselmani,  Fatima',
 'Fynney,  Joseph J',
 'Beesley,  Lawrence',
 'McGowan, . Anna "Annie"',
 'Sloper,  William Thompson',
 'Palsson, . Torborg Danira',
 'Asplund,  Carl Oscar (Selma Augusta Emilia Johansson)',
 'Emir,  Farred Chehab',
 'Fortune,  Charles Alexander',
 'O\'Dwyer, . Ellen "Nellie"',
 'Todoroff,  Lalio',
 'Uruchurtu, Don. Manuel E',
 'Spencer,  W

In [878]:
titanic['Last Name'] = [N.split(",")[0] for N in titanic_name2] 
titanic['Last Name']

0         Braund
1        Cumings
2      Heikkinen
3       Futrelle
4          Allen
         ...    
886     Montvila
887       Graham
888     Johnston
889         Behr
890       Dooley
Name: Last Name, Length: 891, dtype: object

In [879]:
titanic['First Name'] = [N.split(",")[-1:] for N in titanic_name2] 
titanic['First Name']

0                                [  Owen Harris]
1      [  John Bradley (Florence Briggs Thayer)]
2                                     [ . Laina]
3              [  Jacques Heath (Lily May Peel)]
4                              [  William Henry]
                         ...                    
886                               [ Rev. Juozas]
887                          [ . Margaret Edith]
888                [ . Catherine Helen "Carrie"]
889                              [  Karl Howell]
890                                  [  Patrick]
Name: First Name, Length: 891, dtype: object

In [880]:
titanic['Last Name'].value_counts()

Andersson    9
Sage         7
Panula       6
Skoog        6
Carter       6
            ..
Hanna        1
Lewy         1
Mineff       1
Haas         1
Dooley       1
Name: Last Name, Length: 667, dtype: int64

The most common three surnames are Andersson, Sage/Panula/Carter  

## Exercise 2: Binning

When analyzing a continuous variable we often want to break it up into discrete categories (bins). For example, age might be considered as a categorical variable with discrete categories of '18-25', '26-35', etc. Implement a function that takes a dataframe, variable name and cutoff points (e.g. 18, 25, 35, etc.) as inputs and returns a dataframe with a single column, which is binned into desired categories. Run this function on the age column of Titanic dataset. What percentage of passengers falls into each of the age bins?

Tip: You can use `pandas.cut()` method internally.

Function specification:

Function takes 3 arguments:
- `df` - pandas data frame
- `col` - the name of column to be binned
- `cutoffs` - a sequence of cuttoff value to be used for binning

Function returns 1 object:
- `binned_df` - pandas data frame with the selected column binned according to the specification

External libraries:
- `numpy`, `pandas`

In [881]:
# Exercise 2:

# Your code goes here

df = pd.DataFrame(titanic)
df


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Last Name  \
0        0         A/5 21171   7.2500   NaN        S     Braund   
1        0          PC 17599  71.2833   C85        C    Cumings   
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3        0            113803  53.1000  C123        S   Futrelle   
4        0            373450   8.0500   NaN        S      Allen   
..     ...               ...      ...   ...      ...        ...   
886      0            211536  13.0000   NaN        S   Montvila   
887      0            112053  30.0000   B42        S     Graham   
888      2        W./C. 6607  23.4500   NaN        S   Johnston   
889      0            111369  30.0000  C148        C       Behr   
890      0            370376   7.7500   NaN        Q     Dooley   

                                    First Name  
0                              [  Owen Harris]  
1    [  John Bradley (Florence Briggs Thayer)]  
2                                   [ . Laina]  
3            [  Jacques Heath (Lily May Peel)]  
4                            [  William Henry]  
..                                         ...  
886                             [ Rev. Juozas]  
887                        [ . Margaret Edith]  
888              [ . Catherine Helen "Carrie"]  
889                            [  Karl Howell]  
890                                [  Patrick]  

[891 rows x 14 columns]

In [882]:
df["Age_Group"] = 0
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Last Name  \
0        0         A/5 21171   7.2500   NaN        S     Braund   
1        0          PC 17599  71.2833   C85        C    Cumings   
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3        0            113803  53.1000  C123        S   Futrelle   
4        0            373450   8.0500   NaN        S      Allen   
..     ...               ...      ...   ...      ...        ...   
886      0            211536  13.0000   NaN        S   Montvila   
887      0            112053  30.0000   B42        S     Graham   
888      2        W./C. 6607  23.4500   NaN        S   Johnston   
889      0            111369  30.0000  C148        C       Behr   
890      0            370376   7.7500   NaN        Q     Dooley   

                                    First Name  Age_Group  
0                              [  Owen Harris]          0  
1    [  John Bradley (Florence Briggs Thayer)]          0  
2                                   [ . Laina]          0  
3            [  Jacques Heath (Lily May Peel)]          0  
4                            [  William Henry]          0  
..                                         ...        ...  
886                             [ Rev. Juozas]          0  
887                        [ . Margaret Edith]          0  
888              [ . Catherine Helen "Carrie"]          0  
889                            [  Karl Howell]          0  
890                                [  Patrick]          0  

[891 rows x 15 columns]

In [883]:
max(df["Age"]) ## get the max age for grouping 

80.0

In [884]:
age = df["Age"]

In [885]:
df["Age_Group"] = pd.cut(np.array(age),
       4, labels=["1-17", "18-25", "26-35", "36-80"])
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Last Name  \
0        0         A/5 21171   7.2500   NaN        S     Braund   
1        0          PC 17599  71.2833   C85        C    Cumings   
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3        0            113803  53.1000  C123        S   Futrelle   
4        0            373450   8.0500   NaN        S      Allen   
..     ...               ...      ...   ...      ...        ...   
886      0            211536  13.0000   NaN        S   Montvila   
887      0            112053  30.0000   B42        S     Graham   
888      2        W./C. 6607  23.4500   NaN        S   Johnston   
889      0            111369  30.0000  C148        C       Behr   
890      0            370376   7.7500   NaN        Q     Dooley   

                                    First Name Age_Group  
0                              [  Owen Harris]     18-25  
1    [  John Bradley (Florence Briggs Thayer)]     18-25  
2                                   [ . Laina]     18-25  
3            [  Jacques Heath (Lily May Peel)]     18-25  
4                            [  William Henry]     18-25  
..                                         ...       ...  
886                             [ Rev. Juozas]     18-25  
887                        [ . Margaret Edith]      1-17  
888              [ . Catherine Helen "Carrie"]       NaN  
889                            [  Karl Howell]     18-25  
890                                [  Patrick]     18-25  

[891 rows x 15 columns]

Get how to use pandas.cut() here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html 

Now Let's try how to do it with function

In [886]:
def cutoff(df, col, cutoffs):
    df = pd.DataFrame(df)
    col = df[col]
    bins = len(cutoffs) - 1
    df["Age_Group"] = pd.cut(np.array(col),
       bins, labels = ["1-17", "18-25", "26-35", "36-80"]) ## we have found out that the max age is 80 in the previous part
    print(df["Age_Group"])

In [887]:
cutoff(titanic, 'Age', [1, 18, 25, 35, 80])
df

0      18-25
1      18-25
2      18-25
3      18-25
4      18-25
       ...  
886    18-25
887     1-17
888      NaN
889    18-25
890    18-25
Name: Age_Group, Length: 891, dtype: category
Categories (4, object): ['1-17' < '18-25' < '26-35' < '36-80']


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Last Name  \
0        0         A/5 21171   7.2500   NaN        S     Braund   
1        0          PC 17599  71.2833   C85        C    Cumings   
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3        0            113803  53.1000  C123        S   Futrelle   
4        0            373450   8.0500   NaN        S      Allen   
..     ...               ...      ...   ...      ...        ...   
886      0            211536  13.0000   NaN        S   Montvila   
887      0            112053  30.0000   B42        S     Graham   
888      2        W./C. 6607  23.4500   NaN        S   Johnston   
889      0            111369  30.0000  C148        C       Behr   
890      0            370376   7.7500   NaN        Q     Dooley   

                                    First Name Age_Group  
0                              [  Owen Harris]     18-25  
1    [  John Bradley (Florence Briggs Thayer)]     18-25  
2                                   [ . Laina]     18-25  
3            [  Jacques Heath (Lily May Peel)]     18-25  
4                            [  William Henry]     18-25  
..                                         ...       ...  
886                             [ Rev. Juozas]     18-25  
887                        [ . Margaret Edith]      1-17  
888              [ . Catherine Helen "Carrie"]       NaN  
889                            [  Karl Howell]     18-25  
890                                [  Patrick]     18-25  

[891 rows x 15 columns]

Get how to use pandas.cut here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html 

## Exercise 3: Pattern detection

Let us consider the column `Survived`, which contains a binary indicator of whether a given passenger survived the sinking of Titanic. What factors might explain the survival of some passengers but not others? How would you investigate these relationships? Describe your hypotheses and descriptive analyses that can help answering these questions. Perform the outlined descriptive analyses on the dataset.

External libraries:
- `numpy`, `pandas`

In [888]:
# Exercise 3:

# Your text and code go here

Hypothesis: The class of the passenger, Pclass, and gender can explain the variations. The higher class of the pessenger, the higher chance the passenger could survive; female passengers have a higher chance of survivial than male passengers.  

In [889]:
import pandas as pd
import numpy as np

In [906]:
pd.crosstab(titanic['Survived'], titanic["Pclass"], normalize = 'columns')*100

Pclass            1          2          3
Survived                                 
0         37.037037  52.717391  75.763747
1         62.962963  47.282609  24.236253

Survival rate of class 1 is 62.96%; survival rate of class 2 is 47.28%; survival rate of class 3 is 24.24%.
The higher class a passenger is, the more likely the passenger can survive.

In [905]:
pd.crosstab(titanic['Survived'], titanic["Sex"],normalize = 'columns')*100

Sex          female       male
Survived                      
0         25.796178  81.109185
1         74.203822  18.890815

The survival rate of female passenger is 74.20%; the survival rate of male passenger is 18.89%.
Female passengers were more likely to survive than male passengers.

## Exercise 4: Methods

Given below is a simple class definition for class `Person`. At the moment the only method it has is its constructor, that takes person's first name and last name as inputs. Implement 2 additional methods for this class. One would return person's full name, another one would print it out when an object of this class is an argument to `print()` or `str()`. Create an instance of the modified class and call the first method and `print()` function on it.

Method specification:

Method 1 returns 1 object:
- `full_name` - person's full name provided as first name and last name separated by a whitespace

Method 2 returns `None` and prints out the object as:
- `full_name` - person's full name provided as first name and last name separated by a whitespace

External libraries:
- None

In [893]:
class Person(object):
    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name

In [894]:
# Exercise 4:

# Your text and code go here
class Person(object):
    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name
    def full_name(self):
        """Get Person's Full Name"""
        return self.first_name + " " + self.last_name
    def __str__(self):
        """Return None and Print Full Name"""
        return self.first_name + " " + self.last_name 
    



In [895]:
KarlMarx = Person("Karl", "Marx")

In [896]:
print(KarlMarx)

Karl Marx


In [897]:
KarlMarx.first_name

'Karl'

In [898]:
KarlMarx.full_name()

'Karl Marx'

In [899]:
KarlMarx.__str__()

'Karl Marx'

In [900]:
str(KarlMarx)

'Karl Marx'

## Exercise 5: Inheritance

Now create an `Employee` class, which would inherit from `Person`. Add a data attribute for salary. Implement 2 methods, where (1) retrieves employee's salary and (2) gives a salary raise as a percentage of the current salary. Test both functions on an instance of this class.  

Method specification:

Method 1 returns 1 object:
- `salary` - person's annual salary provided as an integer or float

Method 2 takes 1 argument:
- `percentage` - percentage by which employee's salary should be raised

External libraries:
- None

In [901]:
# Exercise 5:

# Your text and code go here


class Employee(Person):
    def __init__(self, first_name, last_name, salary):
        super().__init__(first_name, last_name)
        self.salary = salary
    def salary(self):
        return self.salary 
    def salary_raise(self, percentage):
        self.salary_raise = self.salary * (1 + percentage/100)
        return self.salary_raise 
    def __str__(self):
        return self.first_name + " " + self.last_name + " " + self.salary
            
        

In [902]:
Rossa_Luxembourg = Employee("Rossa", "Luxembourg", 100)

In [903]:
Rossa_Luxembourg.salary

100

In [904]:
Rossa_Luxembourg.salary_raise(50)

150.0

The output is 150.

Note: the salary_raise() method can only be run once. If I click the run button twice, it gets an error. But I can get the output anyway. 

Also, the answer is inspired by: https://www.w3schools.com/python/python_inheritance.asp